benchmarks/src/jmh/java/io/grpc/benchmarks/TransportBenchmark.java

```diff
@@ -177,15 +185,133 @@ public void tearDown() throws Exception {
     }
   }
 
-  private SimpleRequest simpleRequest = SimpleRequest.newBuilder()
+  private static final SimpleRequest UNARY_CALL_1024_REQUEST = SimpleRequest.newBuilder()
       .setResponseSize(1024)
       .setPayload(Payload.newBuilder().setBody(ByteString.copyFrom(new byte[1024])))
       .build();
 
   @Benchmark
   @BenchmarkMode(Mode.SampleTime)
   @OutputTimeUnit(TimeUnit.NANOSECONDS)
-  public SimpleResponse unaryCall1024() {
-    return stub.unaryCall(simpleRequest);
+  public SimpleResponse unaryCall1024Latency() {
+    return stub.unaryCall(UNARY_CALL_1024_REQUEST);
+  }
+
+  private static final SimpleRequest BYTE_THROUGHPUT_REQUEST = SimpleRequest.newBuilder()
+      .setResponseSize(1048576) // must match OperationsPerInvocation
+      .build();
+
+  @Benchmark
+  @BenchmarkMode(Mode.Throughput)
+  @OperationsPerInvocation(1048576) // must match BYTE_THROUGHPUT_REQUEST
+  @Threads(10)
+  public SimpleResponse unaryCallsByteThroughput() {
+    return stub.unaryCall(BYTE_THROUGHPUT_REQUEST);
+  }
+
+  private static Throwable OK_THROWABLE = new RuntimeException("OK");
+
+  @State(Scope.Thread)
+  public static class PingPongStreamState {
+    private final ThreadlessExecutor executor = new ThreadlessExecutor();
+    private StreamObserver<SimpleRequest> requestObserver;
+    private SimpleResponse response;
+    private Throwable status;
+
+    @Setup
+    public void setUp(TransportBenchmark bench) {
+      requestObserver = bench.asyncStub
+          .withExecutor(executor)
+          .streamingCall(new StreamObserver<SimpleResponse>() {
+            @Override public void onNext(SimpleResponse next) {
+              assert response == null;
+              response = next;
+            }
+
+            @Override public void onError(Throwable t) {
+              status = t;
+            }
+
+            @Override public void onCompleted() {
+              status = OK_THROWABLE;
+            }
+          });
+    }
+
+    /** Issues request and waits for response. */
+    public SimpleResponse pingPong(SimpleRequest request) throws InterruptedException {
+      requestObserver.onNext(request);
+      while (true) {
```

Comment:
It seems like waitAndDrain should be at the top.

Reply:
Done.